lib/rules/no-undefined.js

```diff
@@ -21,15 +21,76 @@ module.exports = {
 
     create(context) {
 
+        /**
+         * Checks whether the given reference is 'undefined' or not.
+         *
+         * @param {escope.Reference} reference - The reference to check.
+         * @returns {boolean} `true` if the reference is 'undefined'.
+         */
+        function isUndefinedReference(reference) {
+            const id = reference.identifier;
+
+            return id && id.name === "undefined";
+        }
+
+        /**
+         * Report an invalid "undefined" identifier node.
+         * @param {ASTNode} node The node to report.
+         * @returns {void}
+         */
+        function report(node) {
+            context.report({
+                node,
+                message: "Unexpected use of undefined."
+            });
+        }
+
+        /**
+         * Checks the given scope for references to `undefined` and reports
+         * all references found.
+         * @param {escope.Scope} scope The scope to check.
+         * @returns {void}
+         */
+        function checkScope(scope) {
+            const undefinedVar = scope.variables
+                ? scope.variables.filter(variable => variable.name === "undefined")[0]
+                : null;
+
+            const references = undefinedVar
+                ? undefinedVar.references
+                : [];
+
+            const defs = undefinedVar
+                ? undefinedVar.defs
+                : [];
+
+            // Report non-initializing references (those are covered in defs below)
+            references
+                .filter(ref => !ref.init)
+                .forEach(ref => report(ref.identifier));
+
+            defs.forEach(def => report(def.name));
```

Comment:
I'm hoping to avoid using sets of ASTNodes because that depends on escope maintaining referential equality between their definition names and their reference identifiers in this case. (Unless I'm wrong about how Sets work and they actually do deep compares of their members?)